[Yerin512] 5주차 과제 레스토랑 조회 구현하기

yerin512 opened the pull request:
안녕하세요 트레이너님,
우선 red-green-refactor 로 테스트코드 작성 후, 구현하고 추가로 리팩터링하는 방식으로 진행하기 위해
첫 render시 나오는 화면테스트코드 작성 후,  App 컴포넌트 UI를 구현해보았습니다.

**질문1.**
App 컴포넌트에서 ui화면과 클릭 이벤트, ajax로 데이터까지 불러온 뒤
App 컴포넌트를 관심사를 나눠 분리하고,
마지막으로 redux를 써볼 계획인데... 괜찮은가요? ㅜㅜ
redux를 먼저하는 게 좋을까요? 
사실 redux 자신이 없어서 마지막에 하려고 미뤘습니다...

**질문2**
3주차 과제때 알려주셨던 jest 자동 초기화를 하면 모든 테스트코드 (reducer나 api 테스트코드) mock function을 초기화해주지 않아도 되나요?
그리고 package.json 상위 라인에 
![image](https://user-images.githubusercontent.com/73337811/177153364-46bddbad-d8ee-40f1-98d3-c553eb0dfb46.png)
이렇게 추가해주는게 맞을까요? 문서에는 최상위에 적으라고 적혀있었어요!

yerin512 committed:
chore: 자동으로 mock function을 초기화해주는 jest 설정 추가

package.json에 clearMocks를 추가해주었습니다.

yerin512 committed:
App.jsx의 테스트 코드를 작성했습니다

렌더시에 지역 목록이 보이는 테스트 코드 작성
렌더시에 카테고리 목록이 보이는 테스트 코드 작성

yerin512 committed:
App.jsx 구현

작성한 테스트 코드가 통과될 수 있게 App.jsx의 ui를 구현했습니다.

yerin512 committed:
App.jsx에 버튼 클릭할 때의 테스트코드 추가

handleClick 함수 호출하는 테스트 코드 작성
선택된 버튼에 v표시되는 테스트 코드 작성

yerin512 committed:
App 구현 - click event 추가

작성한 테스트 코드를 따라 handleClick 함수를 작성했습니다.
클릭 된 버튼에 v가 표시되도록 했습니다.

yerin512 committed:
App.test.jsx 수정 - handler 함수 호출 테스트 코드 수정

App에 onClick 핸들러 함수로 handlClick mock fucntion을 props로 주지 않았어서 추가해주었습니다

yerin512 commented:
[87075fb](https://github.com/CodeSoom/react-week5-assignment-1/pull/107/commits/87075fb9fd1c2e93699b7493df025ccaf1e21bc3)

handler 함수 호출 했을때 오류가

![image](https://user-images.githubusercontent.com/73337811/177150543-27934ea1-c0ca-4fee-80b0-5e3942fae6be.png)

이렇게 나는데,  9시 전에 pr을 올리느라 해당 테스트가 오류난 상태로 올렸습니다!

**+** 

한 시간 넘게 이것저것 다른 명칭으로도 바꿔보고, 다른 분의 코드를 봐도
계속 해당 오류가 뜨네요 .ㅜㅜ

`
 expect(jest.fn()).toBeCalled()

    Expected number of calls: >= 1
    Received number of calls:    0
`

1번이상 받는 걸 기대했지만 0번 호출됐다. 왤까요???
 handleClick도 jest.fn() 써서 가짜함수로 선언해놓았고
it 테스트 코드 안에 render도 해주었는데...

yerin512 committed:
refactor: App.jsx 가독성을 위한 코드 수정

구조분해를 이용해 각 배열들 map 함수 속 콜백함수의 가독성을 높임

레스토랑 목록을 예외케이스만 뒀던 걸 레스토랑 목록  map()으로 변경

yerin512 committed:
add: 테스트 코드 작성 때 쓰일 dummy data인 restaurantsInfo를 __fixture__ 폴더에 추가

지역 목록, 카테고리 목록, 레스토랑 목록을 추가함

hannut91 commented:
> App 컴포넌트에서 ui화면과 클릭 이벤트, ajax로 데이터까지 불러온 뒤
App 컴포넌트를 관심사를 나눠 분리하고,
마지막으로 redux를 써볼 계획인데... 괜찮은가요? ㅜㅜ
redux를 먼저하는 게 좋을까요?
사실 redux 자신이 없어서 마지막에 하려고 미뤘습니다...

App 컴포넌트를 먼저 만들고 나서 상황에 따라서 필요할 때 Redux를 도입하는 연습을 해봐도 괜찮고, Redux를 먼저 설정한 후에 하나씩 추가해도 괜찮습니다.

hannut91 commented:
> 3주차 과제때 알려주셨던 jest 자동 초기화를 하면 모든 테스트코드 (reducer나 api 테스트코드) mock function을 초기화해주지 않아도 되나요?

직접 증명해 보세요 ㅎㅎ 가설을 세우고 그 증명할 방법을 찾고 실제로 증명해 보세요

hannut91 commented:
테스트 코드는 우리가 작성한 코드의 예제입니다. 테스트 코드를 먼저 볼게요

```js
const renderApp = () => ((
  render(<App onClick={handleClick} />)
));
```

`<App />`컴포넌트를 사용할 때 `onClick`이벤트가 발생한다는 것을 짐작할 수 있죠. 근데 실제 `<App />`컴포넌트 선언을 볼까요?

```js
export default function App() {
  // ... 생략
}
``` 

Props로 아무런 값도 받고 있지 않죠. `onClick`으로 핸들러 함수를 전달해도 아무것도 하지 않고 있다는 것을 알 수 있겠네요

yerin512 committed:
add: api server data를 가져오기 위한 api.js 파일 추가

기존의 가짜데이터에서 실제 데이터로 변경하기 위해 fetch함수들을 작성했습니다

yerin512 committed:
refactor: 실제 server data로 변경한 후 App.jsx 를 수정

하드코딩 되어있던 data를 api server에게 응답받은 data로 변경했습니다.

선택된 지역과 카테고리가 비어있는지 확인하기 위해 IsEmptyObj 함수를 추가했습니다.

yerin512 committed:
add: 가독성을 위한 공백 추가

yerin512 commented:
안녕하세요 트레이너님!
질문 답변 감사드립니다!

오늘 저녁 개인 사정이 생겨 
App 컴포넌트의 테스트를 수정 먼저 하지 못하고 
api server data를 가져오는 것부터 추가한 뒤 push했습니다.
테스트코드 오류가 너무 많이 나네요!ㅎㅎㅎ 
피드포워드 해주신 부분 참고하여 수정 후 push 하겠습니다@
감사합니다!

hannut91 commented on src/App.jsx:
```diff
@@ -1,5 +1,76 @@
+  }, []);
+
+  useEffect(() => {
+    if (!isEmptyObj(selectedRegion) && !isEmptyObj(selectedCategory)) {
+      loadRestaurants();
+    }
```
선택된 지역과 선택된 카테고리가 기본값이 빈 오브젝트이다보니 빈 오브젝트인지 화인하고 있는 것 같네요. 아예 선택되지 않았을 경우 undefined로 기본값을 가지게 하는 것은 어떨까요?

hannut91 commented on src/App.jsx:
```diff
@@ -1,5 +1,76 @@
+
+  const handleClick = (choiceButton, setSelected) => {
+    setSelected(choiceButton);
+  };
```
choiceButton이 무엇인지 이해하기 어렵네요. 아무래도 선택했을 때 하는 행위가 똑같아서 중복을 제거하고 싶으셨던 것 같네요.

저라면 이렇게 할 것 같아요.

```js
const handleClick = (setter, value) => () =>  {
  setter(value);
};

// ... 생략

<button type="button" onClick={handleClick(setSelectedRegion, region)} />
```

근데 하고 보니 별로 나아진게 없어 보이네요. 그냥 변수명만 바꾸면 될려나 고민이 되네요

hannut91 reviewed (commented):
테스트 짜는게 점점 어려워 보이는데 어떠신가요?

yerin512 committed:
test: App.jsx 비동기적으로 data 불러와서 생긴 오류 수정

우선 setTimeOut으로 통과할 수있게 코드 수정

yerin512 committed:
refactor: 의미가 불분명한 변수명 수정

기존 불분명했던 choiceButton이라 작성한 파라미터 명을 value로 변경

급히 작성하느라 test로 적었던 element를 restaurant로 변경

yerin512 committed:
test: setTimeOut으로 처리했던 비동기 테스트를  async와 await으로 변경

마지막 v처리는 async와 await를 넣기가 애매하여 setTimeOut을 유지

yerin512 commented:
안녕하세요 트레이너님! 

## 코드리뷰 해주신 것들 중 처리하지 못한것 

- jest 자동 초기화 가설을 세우고 증명

```
  const consoleLog = console.log;
  it('spyOn으로 console.log를 mocking하면, console.log는 다른 함수가 된다.', () => {
    jest.spyOn(console, 'log');
    const consoleLogAfterMocking = console.log;

    // 결과는 Success.
    // console.log는 spyOn으로 mocking한 이후 다른 함수가 된다.
    expect(consoleLog).not.toBe(consoleLogAfterMocking);
  });

```

으로 증명 해보려하는데 중간에     jest.clearAllMocks(); 를 추가해도 결과가 똑같네요 ㅜㅜ ;;;
좀 더 해봐야할 거 같습니다.

- 지역, 카테고리가 선택되지 않을 경우 undefined로 변경

그냥 useState() 괄호 안을 useState({}) 에서 useState(undefined)로 변경하려했는데 안 되는군요...! 그래서 props로 받는 loadRestaurants의 파라미터의 초기값을 undefined로 해도 막혀서!ㅠㅠ
**혹시 작은 힌트를 주실 수 있을지..!**

- `테스트 짜는게 점점 어려워 보이는데 어떠신가요?`

벌써부터 막히니까 과제를 이번주안에 완료할 수 있는지 걱정이 좀 되네요!
 비동기여서 바로 버튼 목록이 바로 뜨지않아
setTimeOut으로 했다가 너무 꼼수를 쓰는 거 같아 찝찝하여
jest 문서를 찾아보니 async await을 사용하여 비동기함수 테스크코드를 작성하길래 사용해봤더니
함수가 아닌 경우는 쓰기가 애매하더라고요! ㅜㅜ
따로 fetchData를 mock function으로 만들어야하나?! 고민하고있는데
9시가 다 돼서 우선 push 했습니다. 작은 거 하나하나도 많이 막히네요 갈 길이 태산인데...!

yerin512 committed:
test: (red) Region, RegionContainer 테스트 코드 추가

yerin512 committed:
test: (red) Category, CategoryContainer 테스트 코드 추가

구현 전 간단한 테스트 코드 작성

yerin512 committed:
feat: App컴포넌트에서 RegionContainer 로 관심사 분리

지역목록 상태관리는 RegionContainer에 분리
화면은 Region에 분리

yerin512 committed:
test: (refactor) 지역목록을 render 되는 Region 컴포넌트에 props 추가 및 버튼 명 수정

Region컴포넌트에 가짜 데이터인 regions props로 전달
기존 카테고리 버튼 이름을 적어놨던 걸 지역 이름 중 하나로 변경

yerin512 committed:
feat: App컴포넌트에서 CategoryContainer 로 관심사 분리

카테고리 목록 상태관리는 CategoryContainer에 분리
화면은 Category에 분리

yerin512 committed:
test: (refactor) 카테고리목록을 render 되는 Category 컴포넌트에 props 추가

Category컴포넌트에 가짜 데이터인 categories props로 전달

yerin512 committed:
refactor: App 컴포넌트 분리

RestaurantContainer는 Redux 적용 후 진행 예정

yerin512 committed:
chore: jset clearMocks 자동 설정 추가

기존 package.json에 있던 jest 설정을 지우고 jest.congin.js에 추가

yerin512 committed:
Add: redux 초기 세팅

리덕스를 사용하기 위해 index.jsx를 Provider로 감싸줘
store를 사용할 수 있게 추가해주었고 reducer.js에 필요한 action type을 작성했습니다.

yerin512 committed:
test: (red) reducer 테스트 작성

setCategories, setRegions 액션에 대한 reducer 테스트 코드 작성

yerin512 committed:
refactor: reducer의 테스트코드가 통과될 수 있게 action 생성

setRegions와 setCategories 작성

yerin512 committed:
test: (red) App컴포넌트에서 dispatch 호출을 위한 테스트코드 작성

yerin512 committed:
refactor: (green) App 컴포넌트에서 dispatch 호출 테스트코드 통과를 위한 리팩토링

react-redux 폴더에 useDispatch mock 함수 추가
App.jsx에서 디스패치 호출

yerin512 committed:
refactor: action type인 loadInitialData 작성

fetchRegions와 fetchCategories에서 받아온 데이터들을 setRegions와 setCategories에 넘겨주게끔 작성

yerin512 committed:
add: 테스트코드를 위한 fetchRegions와 fetchCategories mock function 추가

yerin512 committed:
test: (refactor) 기존 지역버튼 클릭시 기대효과를 handle click 함수 호출에서dispatch 호출로 변경

redux를 사용하게 돼서 기존 handleClick 함수 호출보단 dispatch 호출이 의미있어 변경

yerin512 committed:
test: (red) reducer.test.jsx에 지역 선택할 때 reducer 테스트 코드 추가

selectRegion action 타입 추가

yerin512 commented:
안녕하세요 트레이너님!

진행하다가 막히는 부분은....
아샬님의 과제풀이를 보며 진행 중에 있습니다.
어떤 순서대로 커밋을 할지 쪼개면서 하고있기는 하나
너무 많이 commit을 하는 거같아 어느정도로 조절해야할지
좀 고민이 되네요ㅜㅜ!!!

yerin512 committed:
test: (green) App.test.jsx useSelector.mockImplementation 추가

테스트코드 통과를 위해 useSelector mock function 추가

yerin512 committed:
refactor: (green) 지역 선택시 selectRegion 디스패치 호출을 위한 코드 작성

action.js : selectRegion 액션 타입 추가
selectRegion.js : reducers에 selectRegion 상태 변경함수 추가
Region.jsx: onClick 이벤트 함수 수정
RegionContainer.jsx: handleClick 선언 안에 dispatch  호출 삽입

yerin512 committed:
test: (red) 클릭시 dispatch 호출로 변경

기존 handleClick 함수 호출에서 dispatch 호출로 변경

yerin512 committed:
test: (red) selectCategory 액션 타입 호출시 분류가 선택되는 테스트 코드작성

디스패치에 selectCategory 액션타입이 categoryId와 함께 호출되면 분류가 선택됨

yerin512 committed:
refactor: (green) selectCategory 테스트코드 통과를 위해 코드 리팩토링

action.js : selectCategory 함수 추가
CategoryContainer.jsx : categories를 useSelector를 통해 가져오고, dispatch로 selectCategory 호출
Category.jsx : props 값 수정
reducer: selectCategory 타입 추가

yerin512 committed:
test: (green) App.test.jsx useSelector로 가져오는 값에 categries 추가

yerin512 committed:
test: (red) : 레스토랑 목록 테스트코드 작성

지역, 분류가 선택 되었을 시에 레스토랑 목록이 보이는 테스트 코드 작성

yerin512 committed:
test: (red) reducer에 레스토랑 목록이 바뀌는setRestaurants 테스트코드 작성

yerin512 committed:
test: (red) App.test.jsx에 레스토랑을 useSelector를 써서 불러오는 mock 함수 추가

yerin512 committed:
refactor: (green) 레스토랑 목록을 불러오는 테스트 코드 통과를 위한 리팩토링

reduver와 action에 값을 넣어주는setRestaurants와 api 서버에서 레스토랑 데이터를 load하는 loadRestautrants 추가
RestaurantsContainer 생성
테스트 통과를 위한 restaurantsInfo, mocks api 수정

yerin512 committed:
refator: (green) CategoryContainer와 RegionContainer에서 loadRestaurants dispatch로 호출

yerin512 committed:
콘솔로그 정리 및 공백 정리

yerin512 commented:
기능구현과 기본적인 유닛테스트는 통과했으나 커버리지 100% 맞추는 게 관건이네요...!!ㅎㅎ

hannut91 commented on src/Category.jsx:
```diff
@@ -0,0 +1,17 @@
+      {categories.map((category) => (
+        <li key={category.id}>
+          <button type="button" onClick={() => onClick(category.id)}>
+            {selectedCategory?.id !== category.id ? (`${category.name}`) : (`${category.name}(V)`)}
```
조건문에서 대부분의 경우 부정문 보다는 긍정문이 이해하기 쉽습니다. 의도를 드러내려는 것이 아니라면 긍정문을 먼저 사용하는 것을 고려해 보세요

yerin512 replied:
네 명심하겠습니다~!

hannut91 commented on src/RestaurantContainer.jsx:
```diff
@@ -0,0 +1,19 @@
+
+  return (
+    <>
+      {restaurants.length !== 0 ? (
+        <ul>
+          {restaurants.map((restaurant) => (<li key={restaurant.id}>{restaurant.name}</li>))}
+        </ul>
+      ) : (
+        <p>조건에 해당하는 식당이 없습니다.</p>
+      )}
```
이해하기 상당히 어려워요. 레스토랑이 비어있으면 빠르게 종료해서 의도를 드러내면 좋을 것 같아요

```
if (restaurants.length === 0) {
  return <p>조건에 해당하는 식당이 없습니다.</p>;
}

return (
  <>
    <ul>
      {restaurants.map((restaurant) => (<li key={restaurant.id}>{restaurant.name}</li>))}
    </ul>
  </>
)
```

yerin512 replied:
네 다시 한 번 생각해보겠습니다, 감사합니다!

hannut91 reviewed (commented):
아직 비동기 액션에 대한 테스트가 이해하기 어려운 것 같아요. 다음주에는 비동기 액션에 대해서 더 보완하셨으면 좋겠습니다
